code-gen-projects/java/code-gen-demo/src/test/java/org/example/CodeGenTest.java

```diff
@@ -26,22 +26,22 @@ class CodeGenTest {
          StructWithFields s = new StructWithFields();
 
          // set all the fields of `StructWithFields`
-         s.setA("hello");
-         s.setB(12);
-         s.setD(10e2);
+         s.setA(java.util.Optional.of("hello"));
+         s.setB(java.util.Optional.of(12));
+         s.setD(java.util.Optional.of(10e2));
```

Comment:
I think I'd expect a not-required field to be stored as a nullable `T` and the accessors would return `Optional<T>`.
```java
class Foo {
  private T maybeValue;

  void setValue(T value) {
    self.maybeValue = value;
  }
  
  void unsetValue() {
    self.maybeValue = null;
  }

  Optional<T> getValue() {
    return Optional.ofNullable(self.maybeValue);
  }
  
  T expectValue() {
    if (self.maybeValue == null) {
      throw new IllegalStateException();
    }
    return self.maybeValue;
  }
}
```

This is less noisy than having to write `Optional.of(...)` everywhere while still giving you the type safety of `Optional`. Would that work here, or is there a constraint that makes that difficult?

Reply:
Java doesn't include null-ness in the type checking, so it doesn't _necessarily_ give you better type safety. Also, if we're trying to be performant, then allocating a new `Optional` every time someone calls `getValue` is probably not a good idea.

Reply:
@zslayton your solution does make sense.  I had used `Optional<T>`  so that when we perform `writeTo` operation for a given model(e.g. `foo.writeTo(ionWriter)`) we would not have context to whether the value was actually `null` or was it missing while reading some Ion data.

Reply:
> Java doesn't include null-ness in the type checking, so it doesn't necessarily give you better type safety.

It gives you better type safety in the sense that `getValue()` returns you an explicitly maybe-there, maybe-not value instead of a (subtly) nullable value.

> Also, if we're trying to be performant, then allocating a new Optional every time someone calls getValue is probably not a good idea.

Agreed. You could make `private T maybeValue` into `private Optional<T> maybeValue` so it only gets allocated when it's set instead of when it's read.